internal/upgradecheck/header.go

```diff
@@ -0,0 +1,200 @@
+	IsOpenShift      bool   `json:"is_open_shift"`
+}
+
+// generateHeader aggregates data and returns a struct of that data
+// If any errors are encountered, it logs those errors and uses the default values
+func generateHeader(ctx context.Context, cfg *rest.Config, crClient crclient.Client,
+	log logr.Logger, pgoVersion string, isOpenShift bool) *clientUpgradeData {
+
+	return &clientUpgradeData{
+		PGOVersion:       pgoVersion,
+		IsOpenShift:      isOpenShift,
+		DeploymentID:     ensureDeploymentID(ctx, crClient, log),
+		PGOClustersTotal: getManagedClusters(ctx, crClient, log),
+		KubernetesEnv:    getServerVersion(cfg, log),
+	}
+}
+
+// ensureDeploymentID checks if the UUID exists in memory or in a ConfigMap
+// If no UUID exists, ensureDeploymentID creates one and saves it in memory/as a ConfigMap
+// Any errors encountered will be logged and the ID result will be what is in memory
+func ensureDeploymentID(ctx context.Context, crClient crclient.Client, log logr.Logger) string {
+	// If there is no deploymentID in memory, generate one for possible use
+	if deploymentID == "" {
+		deploymentID = string(uuid.NewUUID())
+	}
+
+	cm := manageUpgradeCheckConfigMap(ctx, crClient, log, deploymentID)
+
+	if cm != nil && cm.Data["deployment_id"] != "" {
+		deploymentID = cm.Data["deployment_id"]
+	}
+
+	return deploymentID
+}
+
+// manageUpgradeCheckConfigMap ensures a ConfigMap exists with a UUID
+// If it doesn't exist, this creates it with the in-memory ID
+// If it exists and it has a valid UUID, use that to replace the in-memory ID
+// If it exists but the field is blank or mangled, we update the ConfigMap with the in-memory ID
+func manageUpgradeCheckConfigMap(ctx context.Context, crClient crclient.Client,
+	log logr.Logger, currentID string) *corev1.ConfigMap {
+
+	retrievedCM := &corev1.ConfigMap{}
+	upgradeCheckConfigMapMetadata := naming.UpgradeCheckConfigMap()
+
+	// If no namespace is set, then log this and skip trying to set the UUID in the ConfigMap
+	if upgradeCheckConfigMapMetadata.GetNamespace() == "" {
+		log.V(1).Info("upgrade check issue: namespace not set")
+		return retrievedCM
+	}
+
+	err := crClient.Get(ctx, naming.AsObjectKey(upgradeCheckConfigMapMetadata), retrievedCM)
+	switch {
+	// If we cannot find the ConfigMap, we'll create it
+	case apierrors.IsNotFound(err):
+		break
+	// If we get any other error, log it and use the in-memory deploymentID
+	case err != nil:
+		log.V(1).Info("upgrade check issue: error retrieving configmap",
+			"response", err.Error())
+		return retrievedCM
+	// If the ConfigMap is missing the deploymentID, we'll update it
+	case retrievedCM.Data["deployment_id"] == "":
+		break
+	// If the ConfigMap exists and has a deploymentID, check it for correctness
+	default:
+		_, parseErr := googleuuid.Parse(retrievedCM.Data["deployment_id"])
+		// No error -- the ConfigMap has a valid deploymentID, so return the ConfigMap as is
+		if parseErr == nil {
+			return retrievedCM
```

Comment:
:wrench: Since this function is now using SSA, I would expect the intent to be applied each time this logic is run (instead of returning the current ConfigMap as is done here).

Therefore, in the `default` case here where the ConfigMap exists and has valid ID, I expect that we would simply proceed on to the logic below to generate the intent, specifically using the existing ID we parse here as `deployment_id` in the intent.  That means we will always be applying our full intent each time this is run (which most of the time will not change, but I could see us eventually adding other labels, annotation and/or data to the intent).